File: runtimes/swift/versions/latest/Sources/Runtime/RuntimeResponse.swift
import Foundation

class RuntimeResponse {
    func binary(
        _ bytes: Data,
        statusCode: Int = 200,
        headers: [String: String] = [:]
    ) -> RuntimeOutput {
        RuntimeOutput(
            body: bytes,
            statusCode: statusCode,
            headers: headers
        )
    }

    func text(
        _ body: String,
        statusCode: Int = 200,
        headers: [String: String] = [:]
    ) -> RuntimeOutput {
        if let binaryBody = body.data(using: .utf8) {
            binary(binaryBody, statusCode: statusCode, headers: headers)
        } else {
            binary(Data(), statusCode: statusCode, headers: headers)
        }
    }

    func send(
        _ body: String,
        statusCode: Int = 200,
        headers: [String: String] = [:]
    ) -> RuntimeOutput {
        text(body, statusCode: statusCode, headers: headers)
    }

    func json(
        _ json: [String: Any?],
        statusCode: Int = 200,
        headers: [String: String] = [:]
    ) throws -> RuntimeOutput {
        var outputHeaders = headers

        outputHeaders["content-type"] = "application/json"

        let data = try JSONSerialization.data(
            withJSONObject: json
        )

        let text = String(
            data: data,
            encoding: .utf8
        )

        return self.text(text!, statusCode: statusCode, headers: outputHeaders)
    }

    func empty() -> RuntimeOutput {
        text("", statusCode: 204)
    }

    func redirect(
        _ url: String,
        statusCode: Int = 301,
        headers: [String: String] = [:]
    ) -> RuntimeOutput {
        var outputHeaders = headers

        outputHeaders["location"] = url

        return text("", statusCode: statusCode, headers: outputHeaders)
    }
}

File: runtimes/dotnet/versions/latest/src/RuntimeResponse.cs
using System.Text.Json;

namespace DotNetRuntime
{
    public class RuntimeResponse
    {
        public RuntimeOutput Binary(
            byte[] bytes,
            int statusCode = 200,
            Dictionary<string, string>? headers = null
        )
        {
            return new RuntimeOutput(
                bytes,
                statusCode,
                headers ?? new Dictionary<string, string>()
            );
        }

        public RuntimeOutput Send(
            string body,
            int statusCode = 200,
            Dictionary<string, string>? headers = null
        )
        {
            return Text(body, statusCode, headers ?? new Dictionary<string, string>());
        }

        public RuntimeOutput Text(
            string body,
            int statusCode = 200,
            Dictionary<string, string>? headers = null
        )
        {
            return Binary(
                System.Text.Encoding.UTF8.GetBytes(body),
                statusCode,
                headers ?? new Dictionary<string, string>()
            );
        }

        public RuntimeOutput Json(
            Dictionary<string, object?> json,
            int statusCode = 200,
            Dictionary<string, string>? headers = null
        )
        {
            if (headers == null)
            {
                headers = new Dictionary<string, string>();
            }

            headers.Add("content-type", "application/json");
            return Text(JsonSerializer.Serialize(json), statusCode, headers);
        }

        public RuntimeOutput Empty()
        {
            return Text("", 204, new Dictionary<string, string>());
        }

        public RuntimeOutput Redirect(
            String url,
            int statusCode = 301,
            Dictionary<string, string>? headers = null
        )
        {
            if (headers == null)
            {
                headers = new Dictionary<string, string>();
            }

            headers.Add("location", url);

            return Text("", statusCode, headers);
        }
    }
}
